### PR TITLE
Fixing Master-Detail Forms implementation

### PR DESCRIPTION
**Pull request by nickrandolph**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)
Bug fix

### :arrow_heading_down: What is the current behavior?
Master Details didn't really work using attributes

### :new: What is the new behavior (if this is a feature change)?
Master Details works - 2 scenarios
Define only Master and Detail pages/view models - automatically generates the MasterDetailsPage
Define Root page/view model where the page inherits from MasterDetailsPage

### :boom: Does this PR introduce a breaking change?
No

### :bug: Recommendations for testing
Playground sample now works for Master-Detail

### :thinking: Checklist before submitting

- [ ] All projects build
- [ ] Follows style guide lines ([code style guide](https://github.com/MvvmCross/MvvmCross#code-style-guidelines))
- [ ] Relevant documentation was updated ([docs style guide](https://www.mvvmcross.com/documentation/contribute/mvvmcross-docs-style-guide))
- [ ] Nuspec files were updated (when applicable)
- [ X ] Rebased onto current develop

### CODE REVIEW

File: MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs [hunk 1]
```diff
@@ -199,6 +200,42 @@ public virtual bool CloseCarouselPage(IMvxViewModel viewModel, MvxCarouselPagePr
             }
         }
 
+        public virtual MvxBasePresentationAttribute CreatePresentationAttribute(Type viewModelType, Type viewType)
+        {
+            if (viewType.GetTypeInfo().IsSubclassOf(typeof(ContentPage)))
```
Comment (martijn00) [inline MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs]: We might be able to use this directly again on .NET Standard.

File: MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs [hunk 2]
```diff
@@ -552,6 +607,31 @@ public virtual bool ClosePage(Page rootPage, Page page, MvxPagePresentationAttri
                 return rootPage as TPage;
         }
 
+        public virtual void ReplaceMasterDetailRoot(Page existingMasterDetailPage, Page newRootPage, MvxMasterDetailPagePresentationAttribute masterDetailAttribute)
```
Comment (martijn00) [inline MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs]: I don't understand what this tries to solve. Can you explain the thought behind it?

**Comment by nickrandolph** [comment]
@martijn00, @Cheesebaron hey let me know thoughts on this PR and what needs to be changed/done to get this approved.

File: MvvmCross-Forms/MvvmCross.Forms.Droid/Platform/MvxFormsAndroidSetup.cs [hunk 3]
```diff
@@ -1,28 +1,28 @@
...
-using MvvmCross.Forms.Views;
-using MvvmCross.Forms.Droid.Bindings;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
 
 namespace MvvmCross.Forms.Droid.Platform
 {
-    public abstract class MvxFormsAndroidSetup : MvxAndroidSetup
+    public abstract class MvxFormsAndroidSetup<TForms> : MvxAndroidSetup
```
Comment (martijn00) [inline MvvmCross-Forms/MvvmCross.Forms.Droid/Platform/MvxFormsAndroidSetup.cs]: Why does this need to add a generic? I don't see any value?

File: MvvmCross-Forms/MvvmCross.Forms.Droid/MvvmCross.Forms.Droid.csproj [hunk 4]
```diff
@@ -96,16 +96,16 @@
...
-      <HintPath>..\..\packages\Xamarin.Forms.2.4.0.74863\lib\MonoAndroid10\Xamarin.Forms.Core.dll</HintPath>
+      <HintPath>..\..\packages\Xamarin.Forms.2.5.0.91635\lib\MonoAndroid10\Xamarin.Forms.Core.dll</HintPath>
     </Reference>
     <Reference Include="Xamarin.Forms.Platform, Version=1.0.0.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Xamarin.Forms.2.4.0.74863\lib\MonoAndroid10\Xamarin.Forms.Platform.dll</HintPath>
+      <HintPath>..\..\packages\Xamarin.Forms.2.5.0.91635\lib\MonoAndroid10\Xamarin.Forms.Platform.dll</HintPath>
     </Reference>
     <Reference Include="Xamarin.Forms.Platform.Android, Version=2.0.0.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Xamarin.Forms.2.4.0.74863\lib\MonoAndroid10\Xamarin.Forms.Platform.Android.dll</HintPath>
+      <HintPath>..\..\packages\Xamarin.Forms.2.5.0.91635\lib\MonoAndroid10\Xamarin.Forms.Platform.Android.dll</HintPath>
```
Comment (martijn00) [inline MvvmCross-Forms/MvvmCross.Forms.Droid/MvvmCross.Forms.Droid.csproj]: We should update the nuspec as well if this changes to 2.5

File: MvvmCross-Forms/MvvmCross.Forms.Droid/Platform/MvxFormsAndroidSetup.cs [hunk 5]
```diff
@@ -51,13 +53,16 @@ public MvxFormsApplication FormsApplication
...
-                    _formsApplication = CreateFormsApplication();
+                {
+                    _formsApplication = _formsApplication ?? CreateFormsApplication();
+                }
                 return _formsApplication;
             }
         }
 
-        protected virtual MvxFormsApplication CreateFormsApplication() => new MvxFormsApplication();
+        protected virtual TForms CreateFormsApplication() => new TForms();
```
Comment (martijn00) [inline MvvmCross-Forms/MvvmCross.Forms.Droid/Platform/MvxFormsAndroidSetup.cs]: Maybe we should make this abstract if it gives bugs?

File: MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs [hunk 6]
```diff
@@ -251,18 +298,27 @@ public virtual bool CloseContentPage(IMvxViewModel viewModel, MvxContentPagePres
...
                     masterDetailHost = new MvxMasterDetailPage();
-                    if (attribute.Position == MasterDetailPosition.Master)
-                        masterDetailHost.Master = page;
-                    else
-                        masterDetailHost.Master = new MvxContentPage() { Title = !string.IsNullOrEmpty(attribute.Title) ? attribute.Title : nameof(MvxMasterDetailPage) };
-                    if(attribute.Position == MasterDetailPosition.Detail)
-                        masterDetailHost.Detail = page;
-                    else
-                        masterDetailHost.Detail = new MvxNavigationPage(new MvxContentPage() { Title = !string.IsNullOrEmpty(attribute.Title) ? attribute.Title : nameof(MvxMasterDetailPage) });
+                    //if (attribute.Position == MasterDetailPosition.Master)
```
Comment (martijn00) [inline MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs]: Can you remove any out commented code?

File: MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs [hunk 7]
```diff
@@ -230,14 +277,14 @@ public virtual bool CloseContentPage(IMvxViewModel viewModel, MvxContentPagePres
...
             var page = CreatePage(view, request, attribute);
 
-            if(attribute.Position == MasterDetailPosition.Root)
+            if (attribute.Position == MasterDetailPosition.Root)
             {
                 if (page is MasterDetailPage masterDetailRoot)
                 {
                     if (masterDetailRoot.Master == null)
-                        masterDetailRoot.Master = new MvxContentPage() { Title = !string.IsNullOrEmpty(attribute.Title) ? attribute.Title : nameof(MvxMasterDetailPage)  };
+                        masterDetailRoot.Master = new ContentPage() { Title = !string.IsNullOrEmpty(attribute.Title) ? attribute.Title : nameof(MvxMasterDetailPage) };
```
Comment (martijn00) [inline MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs]: Why change from using `MvxContentPage` to `ContentPage`?

File: MvvmCross/Core/Core/Platform/LogProviders/NullLogProvider.cs [hunk 8]
```diff
@@ -0,0 +1,7 @@
+namespace MvvmCross.Core.Platform.LogProviders
+{
+    internal sealed class NullLogProvider : MvxBaseLogProvider
```
Comment (martijn00) [inline MvvmCross/Core/Core/Platform/LogProviders/NullLogProvider.cs]: This should be removed, and instead set the logger to None.

File: MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs [hunk 9]
```diff
@@ -441,101 +528,208 @@ protected virtual void CloseModalStack(IReadOnlyList<Page> modals)
...
+        //        }
+        //    }
+
+        //    PushOrReplacePage(rootPage, page, attribute);
+        //}
+
         public virtual void PushOrReplacePage(Page rootPage, Page page, MvxPagePresentationAttribute attribute)
         {
-            if (attribute.NoHistory)
+            // Make sure we always have a rootPage
```
Comment (martijn00) [inline MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs]: I'm a bit unsure if this behaviour works in all situations. Can you confirm?

**Comment by martijn00** [comment]
There is definitely some breaking changes in here in terms of behaviour. When testing the playground the modal navigation is different, and the master detail views don't show up with a toolbar with hamburger icon anymore. Also when navigating and coming back the buttons stay disabled. I didn't notice these things before so i think there might be some bugs in here.

File: MvvmCross-Forms/MvvmCross.Forms.Droid/Platform/MvxFormsAndroidSetup.cs (hunk 3 above)
Comment (nickrandolph) [reply MvvmCross-Forms/MvvmCross.Forms.Droid/Platform/MvxFormsAndroidSetup.cs]: I'd suggest I revert to non-generic and that we have a separate generic version of the setup class.

File: MvvmCross-Forms/MvvmCross.Forms.Droid/Platform/MvxFormsAndroidSetup.cs (hunk 5 above)
Comment (nickrandolph) [reply MvvmCross-Forms/MvvmCross.Forms.Droid/Platform/MvxFormsAndroidSetup.cs]: As per my other comment, I think we should revert this change. In the generic setup class the CreateFormsApplication can return a new instance of the generic TForms class

File: MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs (hunk 7 above)
Comment (nickrandolph) [reply MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs]: We were seeing issues where the View Model was being set on the MvxContentPage (passed down from parent to the Master page) which resulted in the OnAppearing method being called multiple times on the master detail root view model

File: MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs (hunk 6 above)
Comment (nickrandolph) [reply MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs]: Will do

File: MvvmCross/Core/Core/Platform/LogProviders/NullLogProvider.cs (hunk 8 above)
Comment (nickrandolph) [reply MvvmCross/Core/Core/Platform/LogProviders/NullLogProvider.cs]: Without this class, the None option for logging will break any code which requests ILogProvider as an injected value. Do you have another alternative to this?

**Comment by nickrandolph** [comment]
A couple of questions/comments:
- Can you be specific on the changes to modal navigation - I tried to make it more consistent as there were cases where it wouldn't show a modal, or when the modal was closed it didn't go back to the master detail
- I'll fix the issue with the icon
- The disabled buttons are actually due to a Delay that's in the Initialize method in one of the view models. Should this prevent navigation until this has completed?

**Pull request merged**